### ISSUE TITLE
[BUG] CSeq decremented erroneously after auth

### ISSUE THREAD

**Issue by bcnewlin** [OpenSIPS/opensips#1969]
**OpenSIPS version you are running**

```
# opensips -V
version: opensips 2.4.6 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 4cdc993c6
main.c compiled on 02:09:03 Feb  6 2020 with gcc 7
```

**Describe the bug**

I’ve run into an issue with CSeq handling after failover where the CSeq number being sent back to the UAC is incorrect (decremented by one). It seems as though there is some flag being set indicating the need for CSeq decrement, which is not being reset when the CSeq is reset on failover.

The timeline is as follows:
- OpenSIPS receives INVITE (CSeq 500)
- OpenSIPS proxies INVITE to dest (CSeq 500)
- OpenSIPS receives 401 response (CSeq 500)
- OpenSIPS sends INVITE with Authentication (CSeq 501)
- OpenSIPS receives 503 response (CSeq 501)
- OpenSIPS sends to new carrier using drouting (CSeq 500)
- OpenSIPS receives 180 response (CSeq 500)
- OpenSIPS proxies 180 response to UAC (CSeq **499**)

I am attaching a PCAP of the error scenario. In the PCAP, OpenSIPS is at 203.0.113.3. The UAS and UAC are both at 203.0.113.2.

**To Reproduce**

I am attaching SIPp UAC/UAS scenarios which will reproduce the error.

We are using the dialog and topology_hiding modules, uac_auth for auth, and drouting for failover after the 503.

**Expected behavior**

The response sent back to upstream should have the same CSeq as the request from upstream.

**Relevant System Logs**

**OS/environment information**
 - Operating System:  Docker amazonlinux:2
 - OpenSIPS installation:  git
 - other relevant information:

**Additional context**

[cseq_error.zip](https://github.com/OpenSIPS/opensips/files/4183539/cseq_error.zip)

**Comment by bogdan-iancu** [comment]
Hey @bcnewlin , let me try to reproduce this

**Comment by bogdan-iancu** [comment]
The problem was the identified and fixed, doing some more testing and finally the backport.

**Comment by bogdan-iancu** [comment]
@bcnewlin , could you please confirm the fix ?

**Comment by bcnewlin** [comment]
@bogdan-iancu I have confirmed this does fix the behavior reported, however it has introduced I believe another bug and also has exposed some behavior I am not sure of.

I am attaching a trace of the same scenario with your fix. The two questions/issues are:
* The BYE is sent from the UAC with CSeq 700, but OpenSIPS is replacing that with CSeq 502 when relaying to the UAS. Is that expected? As a proxy, I assumed OpenSIPS would keep the same CSeq number when proxying requests, even with dialog/topology hiding enabled.
* When the BYE is challenged and failure_route is used to add auth and retry, the CSeq becomes corrupted. It contains both the original UAC CSeq and the replaced OpenSIPS CSeq for the BYE, both incremented by one for the retry.

[tcpdump.pcap.zip](https://github.com/OpenSIPS/opensips/files/4279029/tcpdump.pcap.zip)

**Comment by bogdan-iancu** [comment]
@bcnewlin , let me re-check it

**Comment by bogdan-iancu** [comment]
Long checking :( .....
For the first issue (with the BYE getting changed from 700 to 502), can you please confirm that this is a regression related to the 0200c0e fix? or this issue was present even before /without the fix ?
I'm asking as the fix should impact only the cseq restoring in the replies (sent back to caller), it should have 0 impact on requests handling.

For the second issue, it might be some collision between two changes over the cseq number, so let's try first to sort out the first issue (the unwanted change in the BYE request).

**Comment by bogdan-iancu** [comment]
Hi @bcnewlin, I managed to partially reproduce the scenario (I say partially because so far I used a real UAC -not SIPP- and I couldn't reproduce the cseq jump to 700 in BYE).
Now, on the first point, why the cseq in BYE (before auth) is changed from 700 to 502 - this is done by the dialog module. When the INVITE was auth, the dialog as marked for tracking separate cseq vals on caller and callee side (as the auth may have changed the sequence for cseq from caller to callee) - so opensips cannot act as a simply proxy, it has to re-adjust the cseq to be 100% they are in valid sequence when sending to the callee side (take the case when the auth'ed first branch was successful and opensips will had to manage the +1 offset in cseq for all the sequential requests from caller to callee). Bottom line, because the dialog was marked for "self-cseq management" due the auth process, the BYE get the cseq re-aligned to the next cseq in line (for callee destination), which is 502 (ack was 501). OpenSIPS does not care what was the received cseq, it know what value it must sent out :)
BTW, IMHO the UAC (.4 IP) is broken, as increasing the CSEQ in the 200 OK ACK . In frame .38 the received 200 OK has cseq 500, but the generated ACK  in frame .40 has cseq 501.....

Now, on the second point, the corrupted cseq in BYE. Well, the auth_uac module was intended to be used for auth'ing the initial requests, not sequential ones. What you came across is a conflict between 2 modules trying to change the Cseq in outbound BYE (sent out by .3, in frame 53) after auth - the dialog wants to increase it as for any in-dialog request (as a result of cseq change at INVITE time), so to make it 503; on the other hand, the uac module, when doing auth on BYE, also wants to change the cseq resulting in 701 (the module sees the received cseq value).
I would say you are trying to do a scenario which is not suppose to be supported by OpenSIPS.

**Comment by bcnewlin** [comment]
Excellent analysis and great finds! Thank you!

For the first point, I implemented the jump in Cseq number on BYE (and the UAC bug you found for CSeq on the ACK) as part of testing this issue. My hope was to use SIPp to validate that the CSeq received at the UAS is correct, according to what is sent by the UAC. I may still be able to do that, but I'll need to take this mechanism in OpenSIPS into account rather than assuming the UAC Cseq will always be passed through as is. This is very good to know.

For the second point, this is trickier. SIP allows any request to be challenged/authenticated, including BYE, ACK, CANCEL, etc., and I have worked with several carriers in the past that do authenticate sequential requests like BYE and CANCEL. So I think this is important functionality if it can be supported by fixing this CSeq collision. We have had this BYE authentication in our test suite for quite a while with no issues. If it can't be supported, at least the uac_auth module documentation should be updated to specify it is only supported for initial requests.

**Comment by bogdan-iancu** [comment]
Hi @bcnewlin, based on the code analysis, I'm 99% sure that BYE authentication (as time as the INVITE is also authenticated) will result in a corrupted cseq number (still, not an invalid one). Are you sure about the tests you did ?

**Comment by bcnewlin** [comment]
Prior to this issue we were not verifying the CSeq number in our testing, which is why this seemed like a new problem to me. But other than this CSeq issue, BYE authentication does appear to work, at least in the sense that the header and credentials appear to be offered properly. Our SIPp implementation is verifying the auth credentials in the BYE and they are correct.

**Comment by bogdan-iancu** [comment]
I agree, the auth part (for BYE) works, no doubts, but the cseq is foobar

**Comment by bcnewlin** [comment]
I just wasn't sure when you said this

"the auth_uac module was intended to be used for auth'ing the initial requests, not sequential ones"

whether there was something more than just this CSeq issue which would make the uac_auth module not work with sequential.

**Comment by bogdan-iancu** [comment]
@bcnewlin , should we agree that the original report was addressed by  0200c0e   and close here (if not feel free to re-open :D ). Regarding the second issue mentioned here (auth on BYE), I'm reviewing the patch I sent you off-ticket for being pushed to the public tree.

**Comment by bcnewlin** [comment]
@bogdan-iancu Agreed.
